lib/jwe.js

```diff
@@ -0,0 +1,511 @@
+ * merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to the following
+ * conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS
+ * OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT
+ * OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+ 
+ /* eslint-disable no-new */
+
+'use strict';
+
+var crypto = require('crypto');
+var constants = require('constants');
+const base64url = require('base64url');
+
+var aadutils = require('./aadutils');
+var jwkToPem = require('jwk-to-pem');
+
+/******************************************************************************
+ * utility functions
+ *****************************************************************************/
+
+/**
+ * Create a buffer depends on the version of node
+ *
+ * @param{Buffer/String/Number}  data: buffer, string, or size
+ * @param{String}  encoding: ignored if data is a buffer
+ * @return{Buffer} 
+ */
+var createBuffer = (data, encoding) => {
+  if (!Buffer.isBuffer(data) && typeof data !== 'string' && typeof data !== 'number')
+    throw new Error('in createBuffer, data must be a buffer, string or number');
+
+  if (process.version >= 'v6') {
+    if (typeof data === 'string')
+      return Buffer.from(data, encoding);
+    else if (typeof data === 'number')
+      return Buffer.alloc(data);
+    else
+      return Buffer.from(data);
+  } else {
+    if (typeof data === 'string')
+      return new Buffer(data, encoding);
+    else
+      return new Buffer(data);
+  } 
+};
+
+/**
+ * xor of two number in buffer format
+ *
+ * @param{Buffer}  a: first number
+ * @param{Buffer}  b: second number
+ * @return{Buffer}  a^b
+ */
+var xor = (a, b) => {
+  var la = a.length;
+  var lb = b.length;
+  var c1, c2;
+  if (la > lb) {
+    c1 = a; c2 =b;
+  } else {
+    c2 = a; c1 = b;
+  }
+  var c = createBuffer(c1);
+  for (var i = 1; i <= c2.length; i++)
+    c[c1.length-i] = c[c1.length-i] ^ c2[c2.length-i];
+  return c;
+};
+
+/******************************************************************************
+ * AES key unwrap algorithms
+ *****************************************************************************/
+
+/**
+ * Unwrap wrapped_cek using AES with kek
+ *
+ * @param{String}  algorithm: algorithm
+ * @param{Buffer}  wrapped_cek: wrapped key
+ * @param{Buffer}  kek: key encryption key
+ * @return{Buffer}  cek, the unwrapped key
+ */
+var AESKeyUnWrap = (algorithm, wrapped_cek, kek) => {
```

Comment:
We should perform some parameter checking.
for wrapped_cek n >=3
kek > 128
kek and algorithm match (say, AES256 is specified and kek is 128 bits, what happens?)